towers.rb

```diff
@@ -0,0 +1,234 @@
+  end
+
+
+  # GAME STRUCTURE
+
+  def initialize(tower_height)
+    @tower_height = tower_height
+    set_up_game
+    until game_won?
+      turn
+    end
+    game_ending_sequence
+  end
+
+  def render
+    puts
+    print "L: #{left_stack}\n"
+    print "C: #{center_stack}\n"
+    print "R: #{right_stack}\n"
+    puts
+  end
+
+  def turn
+    render
+    set_origin
+    set_destination
+    move_tile
+    reset_turn_variables
+  end
+
+  def move_tile
+    destination_stack << origin_stack.pop
+  end
+
+  def reset_turn_variables
+    self.take_from = ""
+    self.send_to = ""
+  end
+
+  # PRE-GAME SETUP
+
+  def set_up_game
+    create_model_stack
+    create_tile_stacks
+    create_movement_controls
+    game_welcome_script
+  end
+
+  def create_model_stack
+    @model_stack = (1..tower_height).to_a.reverse
+  end
+
+  def create_tile_stacks
+    @left_stack = model_stack.dup
+    @center_stack = []
+    @right_stack = []
+  end
+
+  def create_movement_controls
+    @origin_stack = []
+    @destination_stack = []
+    @take_from = ""
+    @send_to = ""
+  end
+
+  def game_welcome_script
+    puts "\nWelcome to\nTOWERS OF DAN, OY!\n\n"
+    puts "Enter the tiles you want to move using 'L' for left, 'C' for center and 'R' for right.\nQuit by typing 'quit' at any prompt.\n\n"
+  end
+
+
+  # SETTING ORIGIN
+
+  def set_origin
+    until origin_acceptable?
+      print "Take a tile from stack: \n > "
+      self.take_from = gets.chomp.upcase
+      test_for_quit
+      print_origin_errors
+    end
+    self.origin_stack = stack(take_from)
+  end
+
+  def origin_acceptable?
+    origin_is_valid? && !origin_stack_is_empty?
+  end
+
+  def origin_is_valid?
+    valid_inputs.include? take_from
+  end
+
+  def origin_stack_is_empty?
+    stack(take_from).empty?
+  end
+
+  def print_origin_errors
+    if origin_is_valid?
+      if origin_stack_is_empty?
+        empty_stack_error
+      end
+    else
+      not_a_stack_error
+    end
+  end
+
+
+  # SETTING DESTINATION
+
+  def set_destination
+    until destination_acceptable?
+      print "Move the tile to stack: \n > "
+      self.send_to = gets.chomp.upcase
+      test_for_quit
+      print_destination_errors
+    end
+    self.destination_stack = stack(send_to)
+  end
+
+  def destination_acceptable?
+    destination_is_valid? && destination_can_receive_tile?
+  end
+
+  def destination_is_valid?
+    valid_inputs.include? send_to
+  end
+
+  def destination_can_receive_tile?
+    destination_stack_is_empty? || tile_is_smaller_than_destination?
+  end
+
+  def destination_stack_is_empty?
+    stack(send_to).empty?
+  end
+
+  def tile_is_smaller_than_destination?
+    stack(send_to).last > stack(take_from).last
+  end
+
+  def destination_and_origin_are_same?
+    take_from == send_to
+  end
+
+  def print_destination_errors
+    if destination_is_valid?
+      print_valid_destination_errors
+    else
+      not_a_stack_error
+    end
+  end
+
+  def print_valid_destination_errors
```

Comment:
it might be a good idea to combine some of these valid move checkers into one universal method that checks everything at once